parquet-common/src/main/java/org/apache/parquet/io/InputFile.java

```diff
@@ -41,4 +41,16 @@ public interface InputFile {
    */
   SeekableInputStream newStream() throws IOException;
 
+  /**
+   * Open a new {@link SeekableInputStream} for the underlying data file,
+   * in the range of '[offset, offset + length)'
+   *
+   * @param offset the offset in the file to read from
+   * @param length the total number of bytes to read
+   * @return a new {@link SeekableInputStream} to read the file
+   * @throws IOException if the stream cannot be opened
+   */
+  default SeekableInputStream newStream(long offset, long length) throws IOException {
```

Comment:
If we need to read multiple part of a parquet file (e.g. different row groups, page index, footer, etc.), should we call it multiple times for each individual part?

Reply:
this is what the vectored IO API of hadoop 3.3.5, an API for which we intend to provide in a compatibility library for apps running against hadoop 3.2.x+. https://github.com/apache/hadoop/blob/trunk/hadoop-common-project/hadoop-common/src/site/markdown/filesystem/fsdatainputstream.md#default-void-readvectoredlist-extends-filerange-ranges-intfunctionbytebuffer-allocate

Reply:
problem with length here is it *split length* or *file length*? as with splittable text formats different tasks may get their own split and are allowed to read past it. this is why the fs builder api has two different options for file len and split start/end and we can't use split end as the value for file length.

Reply:
> If we need to read multiple part of a parquet file (e.g. different row groups, page index, footer, etc.), should we call it multiple times for each individual part?

Yea that is the use case we are looking at, ideally we can do parallel downloading for different `Chunk`s in `ParquetFileReader`. Although this PR doesn't cover that part of the changes yet. 

For S3 in particular, we can use this API to only download a range of the file (with `fs.s3a.experimental.input.fadvise` set to `random`), which could give better performance.

> problem with length here is it split length or file length? as with splittable text formats different tasks may get their own split and are allowed to read past it. this is why the fs builder api has two different options for file len and split start/end and we can't use split end as the value for file length.

This is more like "range length". The idea is to break up a row group into smaller ranges (think of column chunks, or consecutive pages in case of column index) which can be processed in parallel.

Reply:
you should go with the hadoop https://issues.apache.org/jira/browse/HADOOP-16202 options; s3a fs now reads them and it lines up abfs/gcs for the same. you can declare split start/end as well as file length so that
* length => client can skip existence probes, they know the file limit
* spilt range: they know to not prefetch past the end of the split, if they prefetch
* read policy: standard set of policies and a parse policy of "csv list of policies -pick the first one you recognise". again, can be used by all the stores